### ISSUE TITLE
and/or not work as expected

### ISSUE THREAD

**Issue by gmazzap** [vimeo/psalm#2474]
According to #98 `and` / `or` should be supported, but they are not fully working.

Consider this: https://psalm.dev/r/196ea5bbc8

Two issues are raised, even if the code works perfectly well: https://3v4l.org/Gjs6D

If I use the ternary operator, like here: https://psalm.dev/r/9ac1485be3 everything is fine.

I know that usage of `or` in the example is not really ideal (even if fully valid), but I used it a very simple way to demonstrate the issue.

In PHP, when there's `{condition} and {statement}`, then `statement` is executed only if `condition` is true (using `or` only if `condition` is false).

This behavior is completely unsupported in Psalm.

Note how even using `&&` and `||` won't help in this case, in fact, it seems that logic operators are taken into consideration inside `if` or other statements like `while`, but now when outside any of those.

**Comment by muglug** [comment]
Does code with this pattern work in the latest release? Since then there have been a number of changes to the master branch (which runs on psalm.dev) which may have broken previously-working code.

**Comment by muglug** [comment]
Nevermind, I just confirmed this is also broken in 3.7.2.

Simpler reproduction, as the only difference between `or` and `||` is operator precedence, which is handled by nikic/php-parser before the code is analysed:

```php
function maybeString(): ?string {
    return rand(0, 10) > 4 ? "test" : null;
}

function test(): string {
    $foo = maybeString();
    ($foo !== null) || ($foo = "");

    return $foo;
}
```

**Comment by muglug** [comment]
I would _really_ discourage you from using this pattern, though – instead just use

```php
function maybeString(): ?string {
    return rand(0, 10) > 4 ? "test" : null;
}

function test(): string {
    $foo = maybeString();
    if ($foo !== null) {
      $foo = "";
    }

    return $foo;
}
```

**Comment by muglug** [comment]
I fixed by just converting these expressions to their if statement equivalents, and evaluating those.

**Comment by gmazzap** [comment]
> I fixed by just converting these expressions to their if statement equivalents, and evaluating those.

Thanks a lot.

> I would really discourage you from using this pattern, though – instead just use

As I said, the code sample was to prove the issue in a simple way, never meant to be real-world code.

That said, `and` / `or` are *very* effective and readable (for _me_, being readability subjective) in some situations like `$foo->doThis() or die();` or simple conditions inside loops without the need to nested `if` that makes code very unreadable to me, and in few other cases.

In any case, even if a pattern might be not "ideal", as long as it is correct, the static analizer should not raise issues. Patterns "niceness" or "desirability" is code-style territory: static analysis is about correctness.

**Comment by muglug** [comment]
> In any case, even if a pattern might be not "ideal", as long as it is correct, the static analizer should not raise issues. Patterns "niceness" or "desirability" is code-style territory: static analysis is about correctness.

Agreed (hence why I marked it as a bug, then fixed)
